src/main/java/org/folio/rest/impl/PutOrderLineByIdHelper.java

```diff
@@ -187,26 +186,39 @@ public boolean isUpdateSuccessful() {
    * @param expectedPiecesWithItem expected Pieces to create with created associated Items records
    * @return void future
    */
-  private CompletableFuture<Void> createMissingPieces(CompositePoLine compPOL, List<Piece> expectedPiecesWithItem) {
-    int itemsSize = expectedPiecesWithItem.size();
-    return searchForExistingPieces(compPOL)
+  private CompletableFuture<Void> createPieces(CompositePoLine compPOL, List<Piece> expectedPiecesWithItem) {
+    int expectedItemsQuantity = expectedPiecesWithItem.size();
+    int expectedPiecesQuantity = expectedItemsQuantity + calculateExpectedQuantityOfPiecesWithoutItemCreation(compPOL, compPOL.getLocations());
+
+    return searchForExistingPieces(compPOL, expectedPiecesQuantity)
       .thenCompose(existingPieces -> {
         List<Piece> piecesToCreate = new ArrayList<>();
+        // For each location collect pieces that need to be created.
         groupLocationsById(compPOL)
           .forEach((locationId, locations) -> {
               List<Piece> filteredExistingPieces = filterByLocationId(existingPieces, locationId);
               List<Piece> filteredExpectedPiecesWithItem = filterByLocationId(expectedPiecesWithItem, locationId);
               piecesToCreate.addAll(collectMissingPiecesWithItem(filteredExpectedPiecesWithItem, filteredExistingPieces));
-              piecesToCreate.addAll(constructMissingPiecesWithoutItem(compPOL, filteredExistingPieces, locations));
+
+              int expectedQuantityOfPiecesWithoutItem = calculateExpectedQuantityOfPiecesWithoutItemCreation(compPOL, locations);
+              int existingQuantityOfPiecesWithoutItem = calculateQuantityOfExistingPiecesWithoutItem(existingPieces);
+              int remainingPiecesQuantity = expectedQuantityOfPiecesWithoutItem - existingQuantityOfPiecesWithoutItem;
+              piecesToCreate.addAll(constructMissingPiecesWithoutItem(compPOL, remainingPiecesQuantity, locationId));
             });
 
         return allOf(piecesToCreate.stream().map(this::createPiece).toArray(CompletableFuture[]::new));
       })
-      .thenAccept(v -> validateItemsCreation(compPOL, itemsSize));
+      .thenAccept(v -> validateItemsCreation(compPOL, expectedItemsQuantity));
   }
 
-  private CompletableFuture<List<Piece>> searchForExistingPieces(CompositePoLine compPOL) {
-    String endpoint = String.format(PIECES_ENDPOINT, compPOL.getId(), lang);
+  /**
+   * Search for pieces which might be already created for the PO line
+   * @param compPOL PO line to retrieve Piece Records for
+   * @param expectedQuantity expected quantity of the pieces for PO Line uuid's from the storage
+   * @return future with list of Pieces
+   */
+  private CompletableFuture<List<Piece>> searchForExistingPieces(CompositePoLine compPOL, int expectedQuantity) {
```

Comment:
The removed `HelperUtils.calculateTotalQuantity` method basically calculated total expected piece quantity. Probably it makes sense to keep it and use here?